File: app/models/product.py
from datetime import datetime, timezone
from enum import Enum
from typing import List, Optional
from beanie import Document, Indexed
from pydantic import Field


class ProductCategory(str, Enum):
    SNACKS = "snacks"
    BEVERAGES = "beverages"
    CAKES = "cakes"
    STATIONERY = "stationery"
    PRINTING = "printing"
    PORTER_5KG = "porter_5kg"


class Product(Document):
    """MongoDB Document model for Quick-Commerce Products & Services."""

    name: Indexed(str)
    category: ProductCategory
    description: Optional[str] = None
    price: float = Field(..., ge=0.0, description="Price in INR")
    unit: str = Field("item", description="Unit of measurement e.g. item, page, pack, trip")
    stock_quantity: int = Field(100, ge=0, description="Available stock quantity")
    is_available: bool = True
    image_url: Optional[str] = None
    tags: List[str] = Field(default_factory=list)
    search_keywords: List[str] = Field(default_factory=list)
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )

    class Settings:
        name = "products"

    def touch(self):
        self.updated_at = datetime.now(timezone.utc)

File: src/models/Product.ts
import mongoose, { Schema, Document } from 'mongoose';

export enum ProductCategory {
  SNACKS = 'snacks',
  BEVERAGES = 'beverages',
  CAKES = 'cakes',
  STATIONERY = 'stationery',
  PRINTING = 'printing',
  PORTER_5KG = 'porter_5kg',
}

export interface IProduct extends Document {
  name: string;
  category: ProductCategory;
  description?: string;
  price: number;
  unit: string;
  stock_quantity: number;
  is_available: boolean;
  image_url?: string;
  tags: string[];
  search_keywords: string[];
  created_at: Date;
  updated_at: Date;
  touch(): void;
}

const ProductSchema = new Schema<IProduct>(
  {
    name: { type: String, required: true, index: true, trim: true },
    category: { type: String, enum: Object.values(ProductCategory), required: true, index: true },
    description: { type: String },
    price: { type: Number, required: true, min: 0 },
    unit: { type: String, default: 'item' },
    stock_quantity: { type: Number, default: 100, min: 0 },
    is_available: { type: Boolean, default: true, index: true },
    image_url: { type: String },
    tags: [{ type: String }],
    search_keywords: [{ type: String }],
  },
  {
    timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' },
  }
);

ProductSchema.index({ category: 1, is_available: 1 });
ProductSchema.index({ is_available: 1, price: 1 });
ProductSchema.index({ name: 'text', description: 'text', search_keywords: 'text', tags: 'text' });

ProductSchema.methods.touch = function () {
  this.updated_at = new Date();
};

export function formatPublicImageUrl(url?: string): string | undefined {
  if (!url) return url;
  const gdriveMatch = url.match(/drive\.google\.com\/(?:file\/d\/|open\?id=|uc\?id=)([a-zA-Z0-9_-]+)/);
  if (gdriveMatch && gdriveMatch[1]) {
    return `https://lh3.googleusercontent.com/d/${gdriveMatch[1]}`;
  }
  return url;
}

ProductSchema.pre('save', function (next) {
  if (this.image_url) {
    this.image_url = formatPublicImageUrl(this.image_url);
  }
  next();
});

export const Product = mongoose.model<IProduct>('Product', ProductSchema, 'products');
